### PR TITLE
Special auto click

### PR DESCRIPTION
**Pull request by Razharas**
Adds automatic mode for guns
Autoclick? What? Never heard
Made specially for gatling lasergun
:cl:
rscadd: some guns now can shoot as long as your mouse button is pressed, moving the mouse changes the target accordingly
/:cl:

### CODE REVIEW

**Comment by TechnoAlchemisto** [comment]
pew pew

**Comment by KazeEspada** [comment]
Fucking russian would make a machine gun

File: code/_onclick/autoclick.dm [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+/mob/living/CanMobAutoclick(object, location, params)
+	var/obj/item/h = get_active_hand()
+	if(h)
+		. = h.CanItemAutoclick(object, location, params)
+
+/obj/item/proc/CanItemAutoclick(object, location, params)
+
+/obj/item/weapon/gun/CanItemAutoclick(object, location, params)
+	. = 1
```
Comment (Core0verload) [inline code/_onclick/autoclick.dm]: Wouldn't that make all the guns full auto?

**Comment by GunHog** [comment]
Delicious laser spam.

File: code/_onclick/autoclick.dm (hunk 1 above)
Comment (Razharas) [inline code/_onclick/autoclick.dm]: It would, i only added the usability while the balance of this is up to discussion

Comment (Core0verload) [inline code/_onclick/autoclick.dm]: Well, it would be !!!!!!!!FUN!!!!!!!! to merge it that way, but no.

File: code/_onclick/autoclick.dm [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/client/MouseDrag(src_object,over_object,src_location,over_location,src_control,over_control,params)
+	if(selected_target)
+		selected_target = over_object
+
+/mob/proc/CanMobAutoclick(object, location, params)
+
+/mob/living/CanMobAutoclick(object, location, params)
+	var/obj/item/h = get_active_hand()
+	if(h)
+		. = h.CanItemAutoclick(object, location, params)
```
Comment (duncathan) [inline code/_onclick/autoclick.dm]: make it `return h.CanItemAutoClick(object, location, params)` instead

File: code/_onclick/autoclick.dm [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+/obj/item/proc/CanItemAutoclick(object, location, params)
+
+/obj/item/weapon/gun
+	var/automatic = 0 //can gun use it, 0 is no, anything above 0 is the delay between clicks in ds
+
+/obj/item/weapon/gun/CanItemAutoclick(object, location, params)
+	. = automatic
```
Comment (duncathan) [inline code/_onclick/autoclick.dm]: Make it `return automati ` instead

File: code/_onclick/click.dm [hunk 4]
```diff
@@ -323,19 +323,29 @@
...
-	icon = 'icons/mob/screen_full.dmi'
-	icon_state = "passage0"
+	icon = 'icons/mob/screen_gen.dmi'
+	icon_state = "click_catcher"
 	plane = CLICKCATCHER_PLANE
 	mouse_opacity = 2
 	screen_loc = "CENTER-7,CENTER-7"
 
+/obj/screen/click_catcher/proc/MakeGreed()
+	var/list/L = list()
```
Comment (duncathan) [inline code/_onclick/click.dm]: Instead of making a new car, use .:
`. = list()`

File: code/_onclick/click.dm [hunk 5]
```diff
@@ -323,19 +323,29 @@
...
+	for(var/i = 0, i<15, i++)
+		for(var/j = 0, j<15, j++)
+			var/obj/screen/click_catcher/CC = new()
+			CC.screen_loc = "NORTH-[i],EAST-[j]"
+			L.Add(CC)
+	return L
```
Comment (duncathan) [inline code/_onclick/click.dm]: then instead of `L.Add(CC)` do `. += CC`
then remove the explicit return

File: code/_onclick/autoclick.dm (hunk 2 above)
Comment (Razharas) [inline code/_onclick/autoclick.dm]: I heard direct returns are slower than default return value setting

File: code/_onclick/autoclick.dm (hunk 3 above)
Comment (Razharas) [inline code/_onclick/autoclick.dm]: Same as above

File: code/_onclick/click.dm (hunk 4 above)
Comment (Razharas) [inline code/_onclick/click.dm]: Ok

File: code/_onclick/click.dm (hunk 5 above)
Comment (Razharas) [inline code/_onclick/click.dm]: Ok

**Comment by KorPhaeron** [comment]
I stole this code

File: code/_onclick/autoclick.dm (hunk 2 above)
Comment (duncathan) [inline code/_onclick/autoclick.dm]: yeah we tested it; turns out we were wrong

Comment (duncathan) [inline code/_onclick/autoclick.dm]: explicit returns are faster for this

Comment (RemieRichards) [inline code/_onclick/autoclick.dm]: well we're not wrong, it just turns out it's RIDICULOUSLY inconsistent, to where there's no point bothering with it or not.

**Pull request closed without merging**